### ISSUE TITLE
Error contacting backend

### ISSUE THREAD

**Issue by NamrataSP** [pwndoc/pwndoc#155]
Hi,
I get an error saying error contacting backend, I've tried adding exceptions, even then there's error rendering the pages. Any help would be appreciated. 
![image](https://user-images.githubusercontent.com/87634546/126443914-2ddc5818-9f11-4ee4-b228-a74614040374.png)
![image](https://user-images.githubusercontent.com/87634546/126443960-a85ea5c1-eb64-455a-a605-ad04c161a350.png)

Also the exceptions are not getting accepted.

**Comment by yeln4ts** [comment]
It seems there is an issue with the connection to mongodb. Can you check the mongo docker image if it's running correcty ?

**Comment by BetaMaxHeadroom** [comment]
Can you please post the results of `sudo docker ps -a`?

**Comment by relaxedricky** [comment]
I've setup a new Debian 10 VM earlier today using the following commands (the same ones I used for my current working instance of pwndoc):

```
sudo apt-get remove docker docker-engine docker.io containerd runc
sudo apt-get update
sudo apt-get install \
    apt-transport-https \
    ca-certificates \
    curl \
    gnupg \
    lsb-release -y
curl -fsSL https://download.docker.com/linux/debian/gpg | sudo gpg --dearmor -o /usr/share/keyrings/docker-archive-keyring.gpg
echo \
  "deb [arch=amd64 signed-by=/usr/share/keyrings/docker-archive-keyring.gpg] https://download.docker.com/linux/debian \
  $(lsb_release -cs) stable" | sudo tee /etc/apt/sources.list.d/docker.list > /dev/null
sudo apt-get update
sudo apt-get install docker-ce docker-ce-cli containerd.io docker-compose git -y
git clone https://github.com/pwndoc/pwndoc.git
cd pwndoc
sudo -g docker docker-compose up -d --build
```

I get the same issue as the op,  looks like mongo-pwndoc is having a few issues:

```
sudo -g docker docker ps -a
CONTAINER ID   IMAGE                     COMMAND                  CREATED       STATUS                            PORTS                                               NAMES
5394b8b25041   yeln4ts/pwndoc:backend    "npm start"              3 hours ago   Up 3 hours                        0.0.0.0:4242->4242/tcp, :::4242->4242/tcp           pwndoc-backend
7805d5ec4333   mongo                     "docker-entrypoint.s…"   3 hours ago   Restarting (132) 30 seconds ago                                                       mongo-pwndoc
29a018e72810   yeln4ts/pwndoc:frontend   "/docker-entrypoint.…"   3 hours ago   Up 3 hours                        80/tcp, 0.0.0.0:8443->8443/tcp, :::8443->8443/tcp   pwndoc-frontend
```

**Comment by yeln4ts** [comment]
I just tried to setup a new instance of pwndoc on a fresh kali VM and Debian 10 VM and I didn't have any issues.
I just did:
```
sudo apt-get update
sudo apt-get install docker docker-compose git
git clone https://github.com/pwndoc/pwndoc.git
cd pwndoc
sudo docker-compose up -d --build
```
I had no issue and all containers were ok
I used this Debian VM: https://www.linuxvmimages.com/images/debian-10/

Can you check the log of the mongodb container to see what's causing the error ?

**Comment by BetaMaxHeadroom** [comment]
Thank you for the response. I think @yeln4ts is correct. It looks like container **7805d5ec4333** is having an issue. If you haven't made any changes, mongo should connect to 127.0.0.1 on port 27017. I have a few suggestions that _might_ be helpful. 

1. Can you please post the results of the "State" tree to command `sudo docker inspect 7805d5ec4333`? There may be an error code and exit code that would be helpful to read into.
2. Set your VM to use bridged network adapter. Using a NAT network adapter in VMWare just caused issues for me. Setting the VM to bridged networking, beyond my comprehension, solved a few issues for me.

**Comment by yeln4ts** [comment]
I think I managed to find the issue (or at least the most probable issue).

First the docker-compose file uses the last version of mongo image, it seems to be ok with a new instance of pwndoc without any database. But there might be some upgrade issues that could corrupt database files. You can view the mongo-pwndoc logs to check for errors.

In my case after upgrading to the latest mongo version the container kept restarting with compatibility issues. After that I specified the mongo version 4.2 in the docker-compose file and it worked again but had some corrupt files that  kept the container restarting every 60s. I then repaired the database files:
```
docker-compose stop
docker run --rm --volumes-from mongo-pwndoc mongo:4.2 unlink "/data/db/mongod.lock"
docker run --rm --volumes-from mongo-pwndoc mongo:4.2 --repair
``` 
And everything was correct again

So for now maybe just specify the mongodb version and try to investigate errors in the log file

**Comment by relaxedricky** [comment]
1.
```
sudo -g docker docker inspect 7805d5ec4333
[
    {
        "Id": "7805d5ec4333c324d8381ee5bc3039b13fcc56dcf3ba1319ca0c58b87a7e5a65",
        "Created": "2021-07-22T12:08:39.019165459Z",
        "Path": "docker-entrypoint.sh",
        "Args": [
            "mongod"
        ],
        "State": {
            "Status": "restarting",
            "Running": true,
            "Paused": false,
            "Restarting": true,
            "OOMKilled": false,
            "Dead": false,
            "Pid": 0,
            "ExitCode": 132,
            "Error": "",
            "StartedAt": "2021-07-22T19:48:24.521761901Z",
            "FinishedAt": "2021-07-22T19:48:24.643623394Z"
        },
        "Image": "sha256:2d38962b1e5e53c2d00efc0f2d45f7508034895b14586363307f4181a42360d4",
        "ResolvConfPath": "/var/lib/docker/containers/7805d5ec4333c324d8381ee5bc3039b13fcc56dcf3ba1319ca0c58b87a7e5a65/resolv.conf",
        "HostnamePath": "/var/lib/docker/containers/7805d5ec4333c324d8381ee5bc3039b13fcc56dcf3ba1319ca0c58b87a7e5a65/hostname",
        "HostsPath": "/var/lib/docker/containers/7805d5ec4333c324d8381ee5bc3039b13fcc56dcf3ba1319ca0c58b87a7e5a65/hosts",
        "LogPath": "/var/lib/docker/containers/7805d5ec4333c324d8381ee5bc3039b13fcc56dcf3ba1319ca0c58b87a7e5a65/7805d5ec4333c324d8381ee5bc3039b13fcc56dcf3ba1319ca0c58b87a7e5a65-json.log",
        "Name": "/mongo-pwndoc",
        "RestartCount": 461,
        "Driver": "overlay2",
        "Platform": "linux",
        "MountLabel": "",
        "ProcessLabel": "",
        "AppArmorProfile": "docker-default",
        "ExecIDs": null,
        "HostConfig": {
            "Binds": [
                "/home/user/pwndoc/backend/mongo-data:/data/db:rw"
            ],
            "ContainerIDFile": "",
            "LogConfig": {
                "Type": "json-file",
                "Config": {}
            },
            "NetworkMode": "pwndoc_backend",
            "PortBindings": {
                "27017/tcp": [
                    {
                        "HostIp": "127.0.0.1",
                        "HostPort": "27017"
                    }
                ]
            },
            "RestartPolicy": {
                "Name": "always",
                "MaximumRetryCount": 0
            },
            "AutoRemove": false,
            "VolumeDriver": "",
            "VolumesFrom": [],
            "CapAdd": null,
            "CapDrop": null,
            "CgroupnsMode": "host",
            "Dns": null,
            "DnsOptions": null,
            "DnsSearch": null,
            "ExtraHosts": null,
            "GroupAdd": null,
            "IpcMode": "shareable",
            "Cgroup": "",
            "Links": null,
            "OomScoreAdj": 0,
            "PidMode": "",
            "Privileged": false,
            "PublishAllPorts": false,
            "ReadonlyRootfs": false,
            "SecurityOpt": null,
            "UTSMode": "",
            "UsernsMode": "",
            "ShmSize": 67108864,
            "Runtime": "runc",
            "ConsoleSize": [
                0,
                0
            ],
            "Isolation": "",
            "CpuShares": 0,
            "Memory": 0,
            "NanoCpus": 0,
            "CgroupParent": "",
            "BlkioWeight": 0,
            "BlkioWeightDevice": null,
            "BlkioDeviceReadBps": null,
            "BlkioDeviceWriteBps": null,
            "BlkioDeviceReadIOps": null,
            "BlkioDeviceWriteIOps": null,
            "CpuPeriod": 0,
            "CpuQuota": 0,
            "CpuRealtimePeriod": 0,
            "CpuRealtimeRuntime": 0,
            "CpusetCpus": "",
            "CpusetMems": "",
            "Devices": null,
            "DeviceCgroupRules": null,
            "DeviceRequests": null,
            "KernelMemory": 0,
            "KernelMemoryTCP": 0,
            "MemoryReservation": 0,
            "MemorySwap": 0,
            "MemorySwappiness": null,
            "OomKillDisable": false,
            "PidsLimit": null,
            "Ulimits": null,
            "CpuCount": 0,
            "CpuPercent": 0,
            "IOMaximumIOps": 0,
            "IOMaximumBandwidth": 0,
            "MaskedPaths": [
                "/proc/asound",
                "/proc/acpi",
                "/proc/kcore",
                "/proc/keys",
                "/proc/latency_stats",
                "/proc/timer_list",
                "/proc/timer_stats",
                "/proc/sched_debug",
                "/proc/scsi",
                "/sys/firmware"
            ],
            "ReadonlyPaths": [
                "/proc/bus",
                "/proc/fs",
                "/proc/irq",
                "/proc/sys",
                "/proc/sysrq-trigger"
            ]
        },
        "GraphDriver": {
            "Data": {
                "LowerDir": "/var/lib/docker/overlay2/c6ab4ec3f36a9ed14af06944d2731bef8be82c664d0ac0f4a40ba836a3267f54-init/diff:/var/lib/docker/overlay2/7404d316d1a217c0586f18c80044f18644f099af4b8e09d69ecc8b3e3cd8c6b7/diff:/var/lib/docker/overlay2/e43fe3a4e1d0cbc8dd5661e70c72ac100d74ba2f275c69f236985af40ca47338/diff:/var/lib/docker/overlay2/61a22b37414b86ead80cd4f38759e2f7f0e29ca2dab587ef141550bf847ec687/diff:/var/lib/docker/overlay2/c15124d590e81d4cc715d9ea7ab784e1f91ce802a32161a20199dd6b2c065c81/diff:/var/lib/docker/overlay2/522b39fec138ad84e53c7d01f52e160cb07d79eb958a18b7dfe81f9152c5e76c/diff:/var/lib/docker/overlay2/7e0d787aa5efb8060b6eb88f6159929a526f1a6e4652259f22d058df5f24fb50/diff:/var/lib/docker/overlay2/541c5fb92d1f16ba93f281e89857e1f67d7ea54ce8f230ce3ef9e432a0243277/diff:/var/lib/docker/overlay2/95557cbb45f3ad5b6f22311c6ab051a7bca92ba6283408a67c1d41086bbc4e04/diff:/var/lib/docker/overlay2/24d36bd845a5051c487b55a69cd6c7334397a431921a055c2fdf7ec405b88eb7/diff:/var/lib/docker/overlay2/0f1172a740b9b0940a0980f1d98dc621cbdd7614a2f33fd969f16e75f3af4421/diff",
                "MergedDir": "/var/lib/docker/overlay2/c6ab4ec3f36a9ed14af06944d2731bef8be82c664d0ac0f4a40ba836a3267f54/merged",
                "UpperDir": "/var/lib/docker/overlay2/c6ab4ec3f36a9ed14af06944d2731bef8be82c664d0ac0f4a40ba836a3267f54/diff",
                "WorkDir": "/var/lib/docker/overlay2/c6ab4ec3f36a9ed14af06944d2731bef8be82c664d0ac0f4a40ba836a3267f54/work"
            },
            "Name": "overlay2"
        },
        "Mounts": [
            {
                "Type": "bind",
                "Source": "/home/user/pwndoc/backend/mongo-data",
                "Destination": "/data/db",
                "Mode": "rw",
                "RW": true,
                "Propagation": "rprivate"
            },
            {
                "Type": "volume",
                "Name": "c6c8739e8488028812722d8de55dec2201442f3873d29fc34a9393101de8851a",
                "Source": "/var/lib/docker/volumes/c6c8739e8488028812722d8de55dec2201442f3873d29fc34a9393101de8851a/_data",
                "Destination": "/data/configdb",
                "Driver": "local",
                "Mode": "",
                "RW": true,
                "Propagation": ""
            }
        ],
        "Config": {
            "Hostname": "7805d5ec4333",
            "Domainname": "",
            "User": "",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "ExposedPorts": {
                "27017/tcp": {}
            },
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": [
                "MONGO_DB:pwndoc",
                "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
                "GOSU_VERSION=1.12",
                "JSYAML_VERSION=3.13.1",
                "MONGO_PACKAGE=mongodb-org",
                "MONGO_REPO=repo.mongodb.org",
                "MONGO_MAJOR=5.0",
                "MONGO_VERSION=5.0.0"
            ],
            "Cmd": [
                "mongod"
            ],
            "Image": "mongo",
            "Volumes": {
                "/data/configdb": {},
                "/data/db": {}
            },
            "WorkingDir": "",
            "Entrypoint": [
                "docker-entrypoint.sh"
            ],
            "OnBuild": null,
            "Labels": {
                "com.docker.compose.config-hash": "05ced09cd23567efe05d56c1ecf7721e1a947a421cfff313fe3c46205516353c",
                "com.docker.compose.container-number": "1",
                "com.docker.compose.oneoff": "False",
                "com.docker.compose.project": "pwndoc",
                "com.docker.compose.service": "mongodb",
                "com.docker.compose.version": "1.21.0"
            }
        },
        "NetworkSettings": {
            "Bridge": "",
            "SandboxID": "fa0fc949f5879d01ce7a7dfaabebea8f7b85720e04127e3fcac6d7ccb97304de",
            "HairpinMode": false,
            "LinkLocalIPv6Address": "",
            "LinkLocalIPv6PrefixLen": 0,
            "Ports": {},
            "SandboxKey": "/var/run/docker/netns/fa0fc949f587",
            "SecondaryIPAddresses": null,
            "SecondaryIPv6Addresses": null,
            "EndpointID": "",
            "Gateway": "",
            "GlobalIPv6Address": "",
            "GlobalIPv6PrefixLen": 0,
            "IPAddress": "",
            "IPPrefixLen": 0,
            "IPv6Gateway": "",
            "MacAddress": "",
            "Networks": {
                "pwndoc_backend": {
                    "IPAMConfig": null,
                    "Links": null,
                    "Aliases": [
                        "mongodb",
                        "7805d5ec4333"
                    ],
                    "NetworkID": "78b15a5b6a3480eca3bf33fa894f9595094fd70dbe685626c34b8a88e04260e9",
                    "EndpointID": "",
                    "Gateway": "",
                    "IPAddress": "",
                    "IPPrefixLen": 0,
                    "IPv6Gateway": "",
                    "GlobalIPv6Address": "",
                    "GlobalIPv6PrefixLen": 0,
                    "MacAddress": "",
                    "DriverOpts": null
                }
            }
        }
    }
]
```

2. The VM is hosted on my proxmox (the same as my working VM)

> First the docker-compose file uses the last version of mongo image, it seems to be ok with a new instance of pwndoc without any database. But there might be some upgrade issues that could corrupt database files. You can view the mongo-pwndoc logs to check for errors.

It's strange that I was seeing the issue on a clean Debian build.

> So for now maybe just specify the mongodb version and try to investigate errors in the log file

I ran your commands and set the version within the docker-compose.yml and the instance runs now without an issue. I will try looking for errors later.

**Comment by NamrataSP** [comment]
Hi @BetaMaxHeadroom,

I found this response for sudo docker ps -a
└─# sudo docker ps -a
CONTAINER ID   IMAGE                               COMMAND                  CREATED        STATUS                            PORTS                                                                                  NAMES
3923ff64d19e   pcf_web                             "/bin/sh -c 'if [ ! …"   18 hours ago   Exited (255) 10 minutes ago       0.0.0.0:5000->5000/tcp, :::5000->5000/tcp                                              pcf_web_1
1e86d8a5fc6d   evait/envizon                       "/entrypoint.sh"         18 hours ago   Exited (1) 18 hours ago                                                                                                  envizon
88a94b0887a2   evait/envizon                       "bundle exec sidekiq…"   18 hours ago   Exited (255) 10 minutes ago                                                                                              envizon_sidekiq_default
84ee732f3424   evait/envizon                       "bundle exec sidekiq…"   18 hours ago   Exited (255) 10 minutes ago                                                                                              envizon_sidekiq_single
f2d8dbf05cdf   redis:alpine                        "docker-entrypoint.s…"   18 hours ago   Exited (255) 10 minutes ago       127.0.0.1:6379->6379/tcp                                                               envizon_redis
7a61b86e94c6   postgres:alpine                     "docker-entrypoint.s…"   18 hours ago   Exited (255) 10 minutes ago       5432/tcp                                                                               envizon_postgres
f1bd183e16ac   selenium/standalone-chrome          "/opt/bin/entry_poin…"   18 hours ago   Exited (255) 10 minutes ago       0.0.0.0:4444->4444/tcp, :::4444->4444/tcp, 0.0.0.0:9515->9515/tcp, :::9515->9515/tcp   envizon_selenium
71e82173a0e4   serpico/serpico                     "/bin/sh -c 'bash sc…"   19 hours ago   Exited (255) 10 minutes ago       0.0.0.0:8443->8443/tcp, :::8443->8443/tcp                                              serpico
c3b44884872a   yeln4ts/pwndoc:backend              "npm start"              21 hours ago   Up 9 minutes                      0.0.0.0:4242->4242/tcp, :::4242->4242/tcp                                              pwndoc-backend
1fe3be9d1c6d   yeln4ts/pwndoc:frontend             "/docker-entrypoint.…"   21 hours ago   Up 9 minutes                      80/tcp, 0.0.0.0:8443->8443/tcp, :::8443->8443/tcp                                      pwndoc-frontend
840287c0738b   mongo                               "docker-entrypoint.s…"   21 hours ago   Restarting (132) 13 seconds ago                                                                                          mongo-pwndoc

**Comment by NamrataSP** [comment]
Hi @yeln4ts,
Adding new version(downloaded newly 4.2.15) to mongo image in the docker-compose.yml file and then rebuilding the instance with `sudo docker-compose up -d`  worked. 

`nano docker-compose.yml`
version: '3'
services:
  mongodb:
    image: mongo:4.2.15
    container_name: mongo-pwndoc
    volumes:
      - ./backend/mongo-data:/data/db
    restart: always
    ports:
      - 127.0.0.1:27017:27017
    environment:
      - MONGO_DB:pwndoc
    networks:
      - backend

  pwndoc-backend:
    build: ./backend
    image: yeln4ts/pwndoc:backend
    container_name: pwndoc-backend
    volumes:
      - ./backend/report-templates:/app/report-templates
      - ./backend/src/config:/app/src/config
    depends_on:
      - mongodb
    restart: always
    ports:
      - 4242:4242
    links:
      - mongodb
    networks:
      - backend

  pwndoc-frontend:
    build: ./frontend
    image: yeln4ts/pwndoc:frontend
    container_name: pwndoc-frontend
    restart: always
    ports:
      - 8443:8443
    networks:
      - backend

networks:
  backend:
      driver: bridge

Pwndoc runs great now.  

Thanks @BetaMaxHeadroom, @yeln4ts @relaxedricky for all the help. 
Will go ahead and close the issue.

**Comment by m6sec** [comment]
The 'latest' tag is a pain and I'm happy that it now uses a defined version tag.

Just wanted to add, that for me I had to pin it to _image: mongo:4.4.6_ as that was the 'latest' version last time I rebuilt the docker image and when it went back to 4.2.15 it had downgrade issues, causing the aforementioned infinite restart loop. 
I suspect this will only impact a small percentage of users, but as far as I can tell 4.4.6 runs stable.
